src/tools/rbd_mirror/LeaderWatcher.cc

```diff
@@ -0,0 +1,712 @@
+}
+
+void LeaderWatcher::handle_post_acquire_leader_lock(int r, Context *on_finish) {
+  dout(20) << "r=" << r << dendl;
+
+  if (r < 0) {
+    if (r == -EAGAIN) {
+      dout(20) << "already locked" << dendl;
+    } else {
+      derr << "error acquiring leader lock: " << cpp_strerror(r) << dendl;
+    }
+    on_finish->complete(r);
+    return;
+  }
+
+  Mutex::Locker locker(m_lock);
+  assert(m_on_finish == nullptr);
+  m_on_finish = on_finish;
+  m_on_finish_r = 0;
+
+  init_status_watcher();
+}
+
+void LeaderWatcher::handle_pre_release_leader_lock(Context *on_finish) {
+  dout(20) << dendl;
+
+  Mutex::Locker locker(m_lock);
+  assert(m_on_finish == nullptr);
+  m_on_finish = on_finish;
+  m_on_finish_r = 0;
+
+  pre_release_leader();
+}
+
+void LeaderWatcher::handle_post_release_leader_lock(int r, Context *on_finish) {
+  dout(20) << "r=" << r << dendl;
+
+  Mutex::Locker locker(m_lock);
+  assert(m_on_finish == nullptr);
+  m_on_finish = on_finish;
+  m_on_finish_r = r;
+
+  notify_lock_released();
+}
+
+void LeaderWatcher::break_leader_lock() {
+  dout(20) << dendl;
+
+  assert(m_lock.is_locked());
+
+  if (m_locker.cookie.empty()) {
+    acquire_leader_lock();
+    return;
+  }
+
+  Context *ctx = create_context_callback<
+    LeaderWatcher, &LeaderWatcher::handle_break_leader_lock>(this);
+
+  m_leader_lock->break_lock(m_locker, true, async(ctx));
+}
+
+void LeaderWatcher::handle_break_leader_lock(int r) {
+  dout(20) << "r=" << r << dendl;
+
+  Mutex::Locker timer_locker(m_threads->timer_lock);
+  Mutex::Locker locker(m_lock);
+
+  if (m_leader_lock->is_shutdown()) {
+    dout(20) << "canceling due to shutdown" << dendl;
+    return;
+  }
+
+  if (r < 0 && r != -ENOENT) {
+    derr << "error beaking leader lock: " << cpp_strerror(r)  << dendl;
+
+    schedule_timer_task("get locker", 1, false, &LeaderWatcher::get_locker);
+    return;
+  }
+
+  acquire_leader_lock();
+}
+
+void LeaderWatcher::get_locker() {
+  dout(20) << dendl;
+
+  assert(m_lock.is_locked());
+
+  std::shared_ptr<librbd::managed_lock::Locker> locker(
+    new librbd::managed_lock::Locker());
+
+  Context *ctx = new FunctionContext(
+    [this, locker](int r) {
+      handle_get_locker(r, *locker);
+    });
+
+  m_leader_lock->get_locker(locker.get(), async(ctx));
+}
+
+void LeaderWatcher::handle_get_locker(int r,
+                                      librbd::managed_lock::Locker& locker) {
+  dout(20) << "r=" << r << dendl;
+
+  Mutex::Locker timer_locker(m_threads->timer_lock);
+  Mutex::Locker mutex_locker(m_lock);
+
+  if (m_leader_lock->is_shutdown()) {
+    dout(20) << "canceling due to shutdown" << dendl;
+    return;
+  }
+
+  if (is_leader(m_lock)) {
+    m_locker = {};
+  } else {
+    if (r == -ENOENT) {
+      acquire_leader_lock();
+    } else {
+      if (r < 0) {
+        derr << "error retrieving leader locker: " << cpp_strerror(r) << dendl;
+      } else {
+        m_locker = locker;
+      }
+
+      schedule_timer_task("break leader", 2, false,
```

Comment:
Two missed heartbeats results in a break lock? This should probably be configurable and be more aligned to 30 seconds like the watch timeout.